src/main/scala/com/twitter/diffy/proxy/RefineHttpHeadersByLabel.scala

```diff
@@ -0,0 +1,54 @@
+package com.twitter.diffy.proxy
+
+import com.twitter.finagle.Filter
+import com.twitter.finagle.http.Request
+import com.twitter.util.Future
+import org.jboss.netty.handler.codec.http.{HttpResponse, HttpHeaders, HttpRequest}
+import scala.collection.JavaConverters._
+
+object RefineHttpHeadersByLabel {
+  val knownLabels: Seq[String] = Seq("primary", "secondary", "candidate")
+
+  def substitute(inclusions: Seq[String], exclusion: Seq[String])(req: HttpRequest) : HttpRequest = {
+
+    def orFn[T](f: T => Boolean, g: T => Boolean): T => Boolean = { t =>  f(t) || g(t) }
+    def constFalse[T](t: T): Boolean = false
+
+    def renameHeaders(headers: HttpHeaders)(nameOld: String, nameNew: String): Unit = {
+      val temp = headers.get(nameOld)
+      headers.remove(nameOld)
+      headers.set(nameNew, temp)
+    }
+
+
+    val removeExclusion  = exclusion.
+      map(_+"_").
+      map( x => (_:String).startsWith(x)).
+      foldLeft(constFalse[String]_)(orFn)
+
+    val reqOut = Request.apply(req)
+    val headers: HttpHeaders = reqOut.headers
+    //k => Option[Function[k => k]]
+    val incomingHeaders = headers.names().asScala.toSet
+
+    val rewrites = for {
+      name <- incomingHeaders
+      inclusion <- inclusions if name.startsWith(inclusion + "_")
+    } yield name->name.substring(inclusion.length+1)
+
+    incomingHeaders.filter( removeExclusion).foreach( headers.remove )
```

Comment:
My reflections - Since the header set is mutable, we should apply all the changes in a single operation.
this makes sense.  It's counter to splitting the filters, unless we just compose aggregates of headerEffects.